drivers/adc/adm1177/iio_adm1177.c

```diff
@@ -0,0 +1,263 @@
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************************************************************/
+#include "iio_adm1177.h"
+#include "adm1177.h"
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+
+static int adm1177_iio_read_raw(void *device, char *buf, uint32_t len,
+				const struct iio_ch_info *channel,
+				intptr_t priv)
+{
+	struct adm1177_iio_dev *iiodev = device;
+	struct adm1177_dev *desc = (struct adm1177_dev *)iiodev->adm1177_dev;
+	int32_t ret, value;
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = conv_voltage;
+
+		break;
+	case IIO_CURRENT:
+		ret = adm1177_read(desc, &conv_voltage, &conv_current);
+		if(ret)
+			return ret;
+
+		value = conv_current;
+
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return iio_format_value(buf, len, IIO_VAL_INT, 1, &value);
+}
+
+static int adm1177_iio_read_scale(void *device, char *buf, uint32_t len,
+				  const struct iio_ch_info *channel,
+				  intptr_t priv)
+{
+	int32_t vals[2];
+	struct adm1177_iio_dev *iio_adm1177;
+	struct adm1177_dev *adm1177;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = device;
+
+	if(!iio_adm1177->adm1177_dev)
+		return -EINVAL;
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	switch(channel->type) {
+	case IIO_VOLTAGE:
+		if(adm1177->vrange == ADM1177_VRANGE_LOW)
+			vals[0] = ADM1177_V_FULLSCALE_0;
+		else
+			vals[0] = ADM1177_V_FULLSCALE_1;
+		vals[1] = 12;
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL_LOG2, 2,
+					vals);
+	case IIO_CURRENT:
+		vals[0] = ADM1177_I_FULLSCALE;
+		vals[1] = 4096 * ADM1177_R_SENSE;
+		return iio_format_value(buf, len, IIO_VAL_FRACTIONAL, 2,
+					vals);
+	default:
+		return -EINVAL;
+	}
+}
+
+static int adm1177_iio_update_channels(void *device, uint32_t mask)
+{
+	struct adm1177_iio_dev *iio_adm1177;
+	uint8_t counter = 0;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = device;
+
+	iio_adm1177->active_channels = mask;
+
+	iio_adm1177->no_of_active_channels = no_os_hweight8(mask);
+
+	return 0;
+}
+
+static int adm1177_iio_read_samples(void *device, int *buff, uint32_t samples)
+{
+	uint16_t conv_voltage;
+	uint16_t conv_current;
+	uint32_t raw_voltage;
+	uint32_t raw_current;
+	int ret;
+	struct adm1177_iio_dev *iio_adm1177;
+	struct adm1177_dev *adm1177;
+
+	if(!device)
+		return -EINVAL;
+
+	iio_adm1177 = device;
+
+	if(!iio_adm1177->adm1177_dev)
+		return -EINVAL;
+
+	adm1177 = iio_adm1177->adm1177_dev;
+
+	for(uint32_t i = 0; i < samples*iio_adm1177->no_of_active_channels;) {
```

Comment:
i am not sure but it seems that the condition is wrong, given that you read 2 channels at the time (with the device driver)

if you have 2 channels enabled, the loop will execute twice, whereas it should've executed only once because your adm1177_read gives you both voltage and current

i think you can drop the for loop

Reply:
Will try it with a while loop.